File: evals/change-pilots/starters/ledger-refunds/typescript/ledger.ts
import { readFileSync } from "node:fs";

// The wire boundary admits only validated operations into the domain.
type Allocation = Readonly<{ invoice: string; amount: number }>;
type IssueInvoice = Readonly<{
  op: "invoice";
  key: string;
  period: number;
  invoice: string;
  total: number;
}>;
type Payment = Readonly<{
  op: "payment";
  key: string;
  period: number;
  allocations: readonly Allocation[];
}>;
type Close = Readonly<{ op: "close"; key: string; through: number }>;
type Mutation = IssueInvoice | Payment | Close;
type ReadOperation = Readonly<{ op: "report" | "audit"; as_of?: number }>;
type Operation = Mutation | ReadOperation;
type Receipt = { event_id: number; replayed: boolean };
type Response =
  | { ok: true; result: unknown }
  | { ok: false; error: { code: string } };

class OperationError extends Error {
  readonly code: string;
  constructor(code = "INVALID_OPERATION") {
    super(code);
    this.code = code;
  }
}
function requireValid(
  condition: unknown,
  code = "INVALID_OPERATION",
): asserts condition {
  if (!condition) throw new OperationError(code);
}
function isObject(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}
function exactFields(
  value: Record<string, unknown>,
  fields: readonly string[],
): boolean {
  return (
    Object.keys(value).length === fields.length &&
    fields.every((key) => Object.hasOwn(value, key))
  );
}
function identifier(value: unknown): value is string {
  return (
    typeof value === "string" && /^[a-z][a-z0-9-]{0,39}(?![\s\S])/.test(value)
  );
}
function integer(
  value: unknown,
  low: number,
  high = Infinity,
): value is number {
  return (
    typeof value === "number" &&
    Number.isInteger(value) &&
    value >= low &&
    value <= high
  );
}
function parseOperation(raw: unknown): Operation {
  requireValid(isObject(raw));
  const kind = raw.op;
  if (kind === "report" || kind === "audit") {
    requireValid(exactFields(raw, ["op"]) || exactFields(raw, ["op", "as_of"]));
    requireValid(!Object.hasOwn(raw, "as_of") || integer(raw.as_of, 0));
    return Object.hasOwn(raw, "as_of")
      ? { op: kind, as_of: raw.as_of as number }
      : { op: kind };
  }
  requireValid(kind === "invoice" || kind === "payment" || kind === "close");
  const fields =
    kind === "invoice"
      ? ["op", "key", "period", "invoice", "total"]
      : kind === "payment"
        ? ["op", "key", "period", "allocations"]
        : ["op", "key", "through"];
  requireValid(exactFields(raw, fields) && identifier(raw.key));
  if (kind === "close") {
    requireValid(integer(raw.through, 1, 1200));
    return { op: kind, key: raw.key, through: raw.through };
  }
  requireValid(integer(raw.period, 1, 1200));
  if (kind === "invoice") {
    requireValid(
      identifier(raw.invoice) && integer(raw.total, 1, 1_000_000_000),
    );
    return {
      op: kind,
      key: raw.key,
      period: raw.period,
      invoice: raw.invoice,
      total: raw.total,
    };
  }
  requireValid(
    Array.isArray(raw.allocations) &&
      raw.allocations.length >= 1 &&
      raw.allocations.length <= 100,
  );
  const seen = new Set<string>();
  const allocations = raw.allocations.map((item: unknown): Allocation => {
    requireValid(isObject(item) && exactFields(item, ["invoice", "amount"]));
    requireValid(
      identifier(item.invoice) && integer(item.amount, 1, 1_000_000_000),
    );
    requireValid(!seen.has(item.invoice));
    seen.add(item.invoice);
    return Object.freeze({ invoice: item.invoice, amount: item.amount });
  });
  return {
    op: kind,
    key: raw.key,
    period: raw.period,
    allocations: Object.freeze(allocations),
  };
}

// Recursive object ordering gives payload equality without changing allocation order.
function canonical(value: unknown): string {
  if (Array.isArray(value)) return `[${value.map(canonical).join(",")}]`;
  if (isObject(value))
    return `{${Object.keys(value)
      .sort()
      .map((key) => `${JSON.stringify(key)}:${canonical(value[key])}`)
      .join(",")}}`;
  return JSON.stringify(value) as string;
}
interface LedgerEvent {
  readonly eventId: number;
  readonly mutation: Mutation;
  readonly payload: Record<string, unknown>;
  readonly fingerprint: string;
}
interface InvoiceBalance {
  total: number;
  paid: number;
}

class Projection {
  closedThrough = 0;
  readonly invoices = new Map<string, InvoiceBalance>();

  apply(mutation: Mutation): void {
    switch (mutation.op) {
      case "invoice":
        this.invoices.set(mutation.invoice, { total: mutation.total, paid: 0 });
        break;
      case "payment":
        for (const allocation of mutation.allocations) {
          this.invoices.get(allocation.invoice)!.paid += allocation.amount;
        }
        break;
      case "close":
        this.closedThrough = mutation.through;
        break;
    }
  }
  report(): unknown {
    const invoices = [...this.invoices.keys()].sort().map((invoice) => {
      const balance = this.invoices.get(invoice)!;
      const due = balance.total - balance.paid;
      return {
        invoice,
        total: balance.total,
        paid: balance.paid,
        refunded: 0,
        due,
        status:
          balance.paid === 0 ? "unpaid" : due === 0 ? "paid" : "partially_paid",
      };
    });
    return {
      closed_through: this.closedThrough,
      invoices,
      cash: invoices.reduce((sum, row) => sum + row.paid, 0),
      outstanding: invoices.reduce((sum, row) => sum + row.due, 0),
    };
  }
}

class Ledger {
  private readonly events: LedgerEvent[] = [];
  private readonly successfulKeys = new Map<string, LedgerEvent>();
  private readonly current = new Projection();

  private validateMutation(mutation: Mutation): void {
    if (mutation.op === "close") {
      requireValid(
        mutation.through > this.current.closedThrough,
        "INVALID_CLOSE",
      );
      return;
    }
    requireValid(mutation.period > this.current.closedThrough, "CLOSED_PERIOD");
    if (mutation.op === "invoice") {
      requireValid(
        !this.current.invoices.has(mutation.invoice),
        "DUPLICATE_INVOICE",
      );
    } else {
      // Validate all references before considering any balance; commit in a later phase.
      requireValid(
        mutation.allocations.every((a) => this.current.invoices.has(a.invoice)),
        "UNKNOWN_INVOICE",
      );
      requireValid(
        mutation.allocations.every((a) => {
          const balance = this.current.invoices.get(a.invoice)!;
          return a.amount <= balance.total - balance.paid;
        }),
        "OVERPAYMENT",
      );
    }
  }
  private read(operation: ReadOperation): unknown {
    const end = operation.as_of ?? this.events.length;
    requireValid(end <= this.events.length, "INVALID_AS_OF");
    const prefix = this.events.slice(0, end);
    if (operation.op === "audit") {
      return {
        events: prefix.map((event) => ({
          event_id: event.eventId,
          operation: structuredClone(event.payload),
        })),
      };
    }
    const snapshot = new Projection();
    for (const event of prefix) snapshot.apply(event.mutation);
    return snapshot.report();
  }
  private mutate(
    mutation: Mutation,
    payload: Record<string, unknown>,
  ): Receipt {
    const fingerprint = canonical(payload);
    const prior = this.successfulKeys.get(mutation.key);
    if (prior) {
      requireValid(prior.fingerprint === fingerprint, "KEY_CONFLICT");
      return { event_id: prior.eventId, replayed: true };
    }
    this.validateMutation(mutation);
    const event: LedgerEvent = Object.freeze({
      eventId: this.events.length + 1,
      mutation: Object.freeze(mutation),
      payload: structuredClone(payload),
      fingerprint,
    });
    this.current.apply(mutation);
    this.events.push(event);
    this.successfulKeys.set(mutation.key, event);
    return { event_id: event.eventId, replayed: false };
  }
  execute(raw: unknown): Response {
    try {
      const operation = parseOperation(raw);
      const result =
        operation.op === "report" || operation.op === "audit"
          ? this.read(operation)
          : this.mutate(operation as Mutation, raw as Record<string, unknown>);
      return { ok: true, result };
    } catch (error) {
      if (error instanceof OperationError)
        return { ok: false, error: { code: error.code } };
      throw error;
    }
  }
}

function run(request: unknown): Response {
  const input = isObject(request) ? request.input : undefined;
  if (
    !isObject(input) ||
    !exactFields(input, ["operations"]) ||
    !Array.isArray(input.operations) ||
    input.operations.length > 500
  ) {
    return { ok: false, error: { code: "INVALID_INPUT" } };
  }
  const ledger = new Ledger();
  return {
    ok: true,
    result: { results: input.operations.map((op) => ledger.execute(op)) },
  };
}

// Preserve the protocol's integer-token distinction (JSON.parse alone turns 1.0 into 1).
const source = readFileSync(0, "utf8").split("\n", 1)[0]!;
const request: unknown = JSON.parse(
  source,
  (_key: string, value: unknown, context?: { source?: string }) => {
    if (typeof value !== "number" || !context?.source) return value;
    if (/[.eE]/.test(context.source)) return NaN;
    // Out-of-range integer tokens still need the correct as_of error. Every valid
    // stored numeric field is bounded, so this sentinel cannot enter an event.
    return Number.isFinite(value) ? value : Math.sign(value) * Number.MAX_VALUE;
  },
);
process.stdout.write(`${JSON.stringify(run(request))}\n`);

File: evals/change-pilots/starters/ledger-refunds/python/ledger.py
"""Append-only baseline ledger: validation, domain events, and prefix projections."""
from __future__ import annotations

from copy import deepcopy
from dataclasses import dataclass, field
import json
import re
import sys
from typing import Any


class OperationError(Exception):
    def __init__(self, code: str) -> None:
        self.code = code


def require(condition: bool, code: str = "INVALID_OPERATION") -> None:
    if not condition:
        raise OperationError(code)


def identifier(value: Any) -> bool:
    return isinstance(value, str) and re.fullmatch(r"[a-z][a-z0-9-]{0,39}", value) is not None


def integer(value: Any, low: int, high: int | None = None) -> bool:
    return type(value) is int and value >= low and (high is None or value <= high)


@dataclass(frozen=True)
class Allocation:
    invoice: str
    amount: int


@dataclass(frozen=True)
class IssueInvoice:
    key: str
    period: int
    invoice: str
    total: int


@dataclass(frozen=True)
class Payment:
    key: str
    period: int
    allocations: tuple[Allocation, ...]


@dataclass(frozen=True)
class Close:
    key: str
    through: int


@dataclass(frozen=True)
class Read:
    kind: str
    as_of: int | None


Mutation = IssueInvoice | Payment | Close
Operation = Mutation | Read


def parse_operation(raw: Any) -> Operation:
    require(type(raw) is dict)
    kind = raw.get("op")
    if kind in ("report", "audit"):
        require(set(raw) in ({"op"}, {"op", "as_of"}))
        require("as_of" not in raw or integer(raw["as_of"], 0))
        return Read(kind, raw.get("as_of"))
    fields = {
        "invoice": {"op", "key", "period", "invoice", "total"},
        "payment": {"op", "key", "period", "allocations"},
        "close": {"op", "key", "through"},
    }
    require(isinstance(kind, str) and kind in fields)
    require(set(raw) == fields[kind])
    require(identifier(raw["key"]))
    if kind == "close":
        require(integer(raw["through"], 1, 1200))
        return Close(raw["key"], raw["through"])
    require(integer(raw["period"], 1, 1200))
    if kind == "invoice":
        require(identifier(raw["invoice"]) and integer(raw["total"], 1, 1_000_000_000))
        return IssueInvoice(raw["key"], raw["period"], raw["invoice"], raw["total"])
    allocations = raw["allocations"]
    require(type(allocations) is list and 1 <= len(allocations) <= 100)
    parsed: list[Allocation] = []
    seen: set[str] = set()
    for item in allocations:
        require(type(item) is dict and set(item) == {"invoice", "amount"})
        require(identifier(item["invoice"]) and integer(item["amount"], 1, 1_000_000_000))
        require(item["invoice"] not in seen)
        seen.add(item["invoice"])
        parsed.append(Allocation(item["invoice"], item["amount"]))
    return Payment(raw["key"], raw["period"], tuple(parsed))


@dataclass(frozen=True)
class Event:
    event_id: int
    mutation: Mutation
    payload: dict[str, Any]


@dataclass
class InvoiceBalance:
    total: int
    paid: int = 0


@dataclass
class Projection:
    closed_through: int = 0
    invoices: dict[str, InvoiceBalance] = field(default_factory=dict)

    def apply(self, mutation: Mutation) -> None:
        if isinstance(mutation, IssueInvoice):
            self.invoices[mutation.invoice] = InvoiceBalance(mutation.total)
        elif isinstance(mutation, Payment):
            for allocation in mutation.allocations:
                self.invoices[allocation.invoice].paid += allocation.amount
        else:
            self.closed_through = mutation.through

    def report(self) -> dict[str, Any]:
        rows = []
        for name, balance in sorted(self.invoices.items()):
            due = balance.total - balance.paid
            rows.append({"invoice": name, "total": balance.total, "paid": balance.paid,
                         "refunded": 0, "due": due,
                         "status": "unpaid" if balance.paid == 0 else "paid" if due == 0 else "partially_paid"})
        return {"closed_through": self.closed_through, "invoices": rows,
                "cash": sum(row["paid"] for row in rows),
                "outstanding": sum(row["due"] for row in rows)}


class Ledger:
    def __init__(self) -> None:
        self.events: list[Event] = []
        self.successful_keys: dict[str, Event] = {}
        self.current = Projection()

    def validate_mutation(self, mutation: Mutation) -> None:
        if isinstance(mutation, Close):
            require(mutation.through > self.current.closed_through, "INVALID_CLOSE")
            return
        require(mutation.period > self.current.closed_through, "CLOSED_PERIOD")
        if isinstance(mutation, IssueInvoice):
            require(mutation.invoice not in self.current.invoices, "DUPLICATE_INVOICE")
        else:
            # Separate passes make unknown invoices win regardless of array order.
            require(all(a.invoice in self.current.invoices for a in mutation.allocations), "UNKNOWN_INVOICE")
            require(all(a.amount <= self.current.invoices[a.invoice].total - self.current.invoices[a.invoice].paid
                        for a in mutation.allocations), "OVERPAYMENT")

    def execute(self, raw: Any) -> dict[str, Any]:
        try:
            operation = parse_operation(raw)
            if isinstance(operation, Read):
                end = len(self.events) if operation.as_of is None else operation.as_of
                require(end <= len(self.events), "INVALID_AS_OF")
                prefix = self.events[:end]
                if operation.kind == "audit":
                    result = {"events": [{"event_id": e.event_id, "operation": deepcopy(e.payload)} for e in prefix]}
                else:
                    snapshot = Projection()
                    for event in prefix:
                        snapshot.apply(event.mutation)
                    result = snapshot.report()
            elif operation.key in self.successful_keys:
                original = self.successful_keys[operation.key]
                require(raw == original.payload, "KEY_CONFLICT")
                result = {"event_id": original.event_id, "replayed": True}
            else:
                self.validate_mutation(operation)
                event = Event(len(self.events) + 1, operation, deepcopy(raw))
                # Commit only after every validation pass has succeeded.
                self.current.apply(operation)
                self.events.append(event)
                self.successful_keys[operation.key] = event
                result = {"event_id": event.event_id, "replayed": False}
            return {"ok": True, "result": result}
        except OperationError as error:
            return {"ok": False, "error": {"code": error.code}}


def run(request: Any) -> dict[str, Any]:
    task_input = request.get("input") if type(request) is dict else None
    if (type(task_input) is not dict or set(task_input) != {"operations"}
            or type(task_input["operations"]) is not list or len(task_input["operations"]) > 500):
        return {"ok": False, "error": {"code": "INVALID_INPUT"}}
    ledger = Ledger()
    return {"ok": True, "result": {"results": [ledger.execute(op) for op in task_input["operations"]]}}


if __name__ == "__main__":
    print(json.dumps(run(json.loads(sys.stdin.readline())), separators=(",", ":")))
